internal/charmstore/search.go

```diff
@@ -76,36 +76,38 @@ func (s *Store) UpdateSearchAsync(r *router.ResolvedURL) {
 	})
 }
 
-// UpdateSearch updates the search record for the entity reference r.
-// The search index only includes the latest revision of each entity so
-// the latest revision of the charm specified by r will be indexed.
+// UpdateSearch updates the search record for the entity reference r. The
+// search index only includes the latest stable revision of each entity
+// so the latest stable revision of the charm specified by r will be
+// indexed.
 func (s *Store) UpdateSearch(r *router.ResolvedURL) error {
 	if s.ES == nil || s.ES.Database == nil {
 		return nil
 	}
-	if r.URL.Series != "" && !series.Series[r.URL.Series].SearchIndex {
-		return nil
+	// For multi-series charms update the whole base URL.
+	if r.URL.Series == "" {
+		return s.UpdateSearchBaseURL(r.PreferredURL())
 	}
 
-	var query *mgo.Query
-	query = s.DB.Entities().Find(bson.D{
-		{"user", r.URL.User},
-		{"name", r.URL.Name},
-		{"series", r.URL.Series},
-	}).Sort("-revision")
-	var entity mongodoc.Entity
-	if err := query.One(&entity); err != nil {
-		if err == mgo.ErrNotFound {
-			return errgo.WithCausef(nil, params.ErrNotFound, "entity not found %s", r)
-		}
-		return errgo.Notef(err, "cannot get %s", r)
+	if !series.Series[r.URL.Series].SearchIndex {
+		return nil
+	}
+	baseEntity, err := s.FindBaseEntity(r.PreferredURL(), nil)
+	if err != nil {
+		return errgo.NoteMask(err, fmt.Sprintf("cannot update search record for %q", r.PreferredURL()), errgo.Is(params.ErrNotFound))
+	}
+	series := r.URL.Series
+	entityURL := baseEntity.StableSeries[series]
+	if entityURL == nil {
+		// There is no stable version of the entity to index.
+		return nil
 	}
-	baseEntity, err := s.FindBaseEntity(entity.BaseURL, nil)
+	entity, err := s.FindEntity(&router.ResolvedURL{URL: *entityURL}, nil)
```

Comment:
This is much better, thanks, and actually corresponds to the correct lookup semantics too.